File: pod_api/bin/app/services/itunes_service.dart
import 'dart:convert';

import '../models/charts_response.dart';
import '../models/lookup_response.dart';
import 'base_service.dart';
import 'package:meta/meta.dart';

class ItunesService {
  BaseService service;

  ItunesService(this.service);

  /// Get top podcasts from the itunes search api.
  ///
  /// Parameters:
  /// - `String language` (default `de`)
  /// - `int limit` (default `30`, max: `200`)
  /// - `bool explicit` (default `true`)
  /// - `String genre` (default `null`, would be the genre id)
  ///
  /// Returns:
  /// - `ChartsResponse`
  Future<ChartsResponse> getTopPodcasts({
    String language = 'de',
    int limit = 30,
    bool explicit = true,
    String genre = null,
    bool withLookup = true,
  }) async {
    final String url = genre == null
        ? '${service.itunesBaseUrl}$language/rss/toppodcasts/limit=$limit/explicit=$explicit/xml'
        : '${service.itunesBaseUrl}$language/rss/toppodcasts/limit=$limit/genre=$genre/explicit=$explicit/xml';

    final xmlResponse = await service.dio.get<String>(url);
    var itunesCharts = ChartsResponse.fromXml(xmlResponse.data);

    if (withLookup) {
      for (var item in itunesCharts.entries) {
        final lookup = await getLookup(id: item.id);
        item.lookup = lookup.results.single;
      }
    }

    return itunesCharts;
  }

  Future<LookupResponse> getLookup({@required String id}) async {
    final String url = '${service.itunesBaseUrl}lookup?id=$id';
    final lookupData = await service.dio.get<String>(url);
    return LookupResponse.fromMap(json.decode(lookupData.data));
  }

  Future<LookupResponse> search({String seachTerm, bool explicit, String language = 'de'}) async {
    final String url =
        '${service.itunesBaseUrl}search?term=$seachTerm&country=$language&media=podcast&explicit=$explicit';
    final searchData = await service.dio.get<String>(url);
    return LookupResponse.fromMap(json.decode(searchData.data));
  }
}

File: podapi_ts/src/Services/ITunesService.ts
import * as xmlParser from "fast-xml-parser";
import { ObjectMapper } from "json-object-mapper";
import * as request from "request-promise";
import { ChartResponse } from "../Models/ChartResponse";
import FeedResponseDTO from "../Models/DTOs/FeedResponseDTO";
import LookupResponseDTO from "../Models/DTOs/LookupResponseDTO";
import { LookupResponse } from "../Models/LookupResponse";
import BaseService from "./BaseService";

/**
 * ### ITunesService
 * Fetching Podcasts and Lookups from the iTunes Api
 */
class ITunesService extends BaseService {
  /**
   *
   * @param langauge Language Code, eg. 'us', 'de' and so on. Default value `'de'`.
   * @param limit Limit of entries. Maximum 200, minimum 1. Default value `30`.
   * @param explicit Explicit entries allowed. Default value `true`
   * @param genre Get entries for genre with genre id. Default value `null`.
   * @param withLookup Get Lookup for more details and direct link to rss file. Default value `false`.
   *
   * @return Returning a Promise with [[ChartResponse]]
   */
  public async getTopPodcasts(
    langauge: string = "de",
    limit: number = 30,
    explicit: boolean = true,
    genre?: string,
    withLookup: boolean = true
  ): Promise<ChartResponse> {
    const url =
      genre == null
        ? `${this.baseUrl}${langauge}/rss/toppodcasts/limit=${limit}/explicit=${explicit}/xml`
        : `${this.baseUrl}${langauge}/rss/toppodcasts/limit=${limit}/genre=${genre}/explicit=${explicit}/xml`;

    return new Promise<ChartResponse>((resolve, reject) => {
      request
        .get(url)
        .then((response: string) => {
          const jsonFromXml = xmlParser.parse(response, {
            ignoreAttributes: false,
            ignoreNameSpace: true
          });

          const feedResponse = ObjectMapper.deserialize(
            FeedResponseDTO,
            jsonFromXml
          );

          return new ChartResponse(feedResponse);
        })
        .then(async chartResponse => {
          if (withLookup) {
            // tslint:disable-next-line: prefer-array-literal
            const promises: Array<Promise<LookupResponse>> = [];

            chartResponse.entries.forEach(entry => {
              promises.push(this.getLookup(entry.id));
            });

            const lookups = await Promise.all(promises);

            chartResponse.entries.forEach(entry => {
              entry.lookup = lookups.find(
                lu => lu.results[0].trackId === Number(entry.id)
              ).results[0];
            });

            return chartResponse;
          } else {
            return chartResponse;
          }
        })
        .then(lookupedResponse => {
          resolve(lookupedResponse);
        })
        .catch(error => {
          return reject(`error: ${error}`);
        });
    });
  }

  public async getLookup(id: string): Promise<LookupResponse> {
    const url = `${this.baseUrl}lookup?id=${id}`;
    return new Promise<LookupResponse>((resolve, reject) => {
      request.get(url, (error, response, body) => {
        if (error != null || response.statusCode === 400) {
          return reject(`error: ${error ?? "Status: " + response.statusCode}`);
        }

        const lookupResponse = ObjectMapper.deserialize(
          LookupResponseDTO,
          JSON.parse(body)
        );

        return resolve(new LookupResponse(lookupResponse));
      });
    });
  }

  public async search(
    searchTerm: string,
    country: string,
    explicit: boolean
  ): Promise<LookupResponse> {
    const url = `${this.baseUrl}search?term=${searchTerm}&country=${country}&media=podcast&explicit=${explicit}`;
    return new Promise<LookupResponse>((resolve, reject) => {
      request.get(url, (error, response, body) => {
        if (error != null || response.statusCode === 400) {
          return reject(`error: ${error ?? "Status: " + response.statusCode}`);
        }

        const lookupResponse = ObjectMapper.deserialize(
          LookupResponseDTO,
          JSON.parse(body)
        );

        return resolve(new LookupResponse(lookupResponse));
      });
    });
  }
}

export default ITunesService;
